docs/node/mainnet/e-upgrade.md

```diff
@@ -100,3 +100,312 @@ upgrade. For now, please take a look at the related [Testnet upgrade].
 :::
 
 [Testnet upgrade]: ../testnet/upgrade-log.md
+
+### Voting
+
+On XXX, an upgrade proposal will be proposed which (if accepted) will schedule
+the upgrade on epoch **XXX**.
+
+For optimal voting experience, we recommend using the [Oasis CLI].
+Follow these steps to cast your vote:
+1. [Import your keys into the wallet]
+2. [Cast your vote]
+
+### Upgrade Instructions
+
+The following steps should be performed only after the network has reached the
+upgrade epoch and has halted:
+
+1. Wait for the network to reach the upgrade epoch and halt. The node
+    will automatically stop without any action required on your part. After the
+    network has halted, proceed to the next steps.
+
+2. Download the Mainnet genesis file published in the
+   [Mainnet XXX release].
+
+:::info
+
+Mainnet state at epoch **XXX** will be exported and migrated to a 23.0.x
+compatible genesis file.
+
+The new genesis file will be published on the above link soon after reaching the
+upgrade epoch.
+
+:::
+
+3. Verify the provided Mainnet upgrade genesis file by comparing it to the
+   local network state dump.
+
+   Find the `genesis-mainnet-XXX.json` file in the `exports` subdirectory in
+   your data dir (e.g. `/node/`, `/srv/oasis/node/`)  and run `sha256sum` on it.
+   Afterwards, compare it with the hash that we will share on the
+   `#node-operators` Discord channel.
+
+   The state changes are described in the [State Changes](#state-changes)
+   section below.
+
+4. Replace the old genesis file with the new Mainnet genesis file.
+
+5. Ensure your node will remain stopped by disabling auto-starting via your
+   process manager (e.g., [systemd] or [Supervisor]).
+
+6. Back up the entire data directory of your node. Verify that the backup
+includes the following folders:
+- for consensus: `tendermint/abci-state` and `tendermint/data`
+- for runtimes: `runtimes/*/mkvs_storage.badger.db` and
+`runtimes/*/worker-local-storage.badger.db`
+
+7. [Wipe state]. This must be performed _before_ replacing the Oasis Node
+   binary. In case you are upgrading ParaTimes/runtimes ensure you read the
+   following section:
+
+:::danger
+
+State of ParaTimes/runtimes is not affected by this upgrade and MUST NOT be
+wiped. Wiping state for confidential ParaTimes will prevent your compute or
+key manager node from transitioning to the new network.
+
+To safely wipe the blockchain state on a runtime while preserving the runtime
+state, follow these steps:
+1. Go into your node's data directory (e.g., `/serverdir/node`, `/node/`,
+`/srv/oasis/node/`)
+2. **Dry run:** initiate a dry run to preview which files will be deleted by
+running the following command: 
+```
+oasis-node unsafe-reset --dry_run
```

Comment:
Have you tested that this works? If I don't specify the data directory I get:
```
level=error ts=2023-10-20T07:42:46.231628389Z caller=unsafe_reset.go:71 module=cmd/unsafe-reset msg="data directory must be set"
```

Reply:
I did not test this. Don't want to test this on my node, so it doesn't get accidentally wiped. I just assumed it would work. I changed the instructions now, to specify the data directory.